### ISSUE TITLE
`#[warn(unused_braces)]` false positive for macro that accepts types and expressions and forwards them to a function

### ISSUE THREAD

**Issue by ZenPZero** [rust-lang/rust#158747]
### Code

```Rust
macro_rules! type_or_expr {
    ($x:ty) => {
        identity(stringify!($x))
    };
    
    ($x:expr) => {
        identity($x)
    }
}

fn main() {
    println!("{}", type_or_expr!({ format!("{}", 1) }));
    println!("{}", type_or_expr!(format!("{}", 1) ));
}

fn identity<T>(x: T) -> T { x }
```

### Current output

```Shell
warning: unnecessary braces around function argument
  --> src/main.rs:12:34
   |
12 |     println!("{}", type_or_expr!({ format!("{}", 1) }));
   |                                  ^^                ^^
   |
   = note: `#[warn(unused_braces)]` (part of `#[warn(unused)]`) on by default
help: remove these braces
   |
12 -     println!("{}", type_or_expr!({ format!("{}", 1) }));
12 +     println!("{}", type_or_expr!(format!("{}", 1) ));
   |

warning: `playground` (bin "playground") generated 1 warning (run `cargo fix --bin "playground" -p playground` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.67s
     Running `target/debug/playground`
```

### Desired output

```Shell
Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.67s
     Running `target/debug/playground`
```

### Rationale and extra context

Removing the braces changes which arm of the macro matches and thus this is a false positive.
Output:
```
1
format!("{}", 1)
```

Also, the suggested change has a formatting issue: the space before the closing brace is not collapsed.
```diff
warning: unnecessary braces around function argument
  --> src/main.rs:12:34
   |
12 |     println!("{}", type_or_expr!({ format!("{}", 1) }));
   |                                  ^^                ^^
   |
   = note: `#[warn(unused_braces)]` (part of `#[warn(unused)]`) on by default
help: remove these braces
   |
12 -     println!("{}", type_or_expr!({ format!("{}", 1) }));
-12 +     println!("{}", type_or_expr!(format!("{}", 1) ));
+12 +     println!("{}", type_or_expr!(format!("{}", 1)));
   |
```
If you remove the first arm of the macro and remove the second invocation, the formatting issue persists, despite it no longer being a false positive

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.96.0 (ac68faa20 2026-05-25)
binary: rustc
commit-hash: ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96
commit-date: 2026-05-25
host: x86_64-unknown-linux-gnu
release: 1.96.0
LLVM version: 22.1.2
```

### Anything else?

Compiler output given from playground, but reproducible locally with the supplied version
